difftest/dpi_t1rocketemu/src/bus/disp.rs

```diff
@@ -0,0 +1,149 @@
+use super::ShadowDevice;
+use std::fs::File;
+use std::io::BufWriter;
+use std::path::PathBuf;
+
+/// Device memory layout:
+/// `0x000_0000 - 0x1FF_0000` addressable frame buffer memory
+/// `0x1FF_0000 - 0x200_0000` control registers
+/// `0x1FF_0000`: read as frame counter, write to flush frame buffer as png
+/// `0x1FF_0004`: output dimensions
+///     0x04-06 16bits: OUTPUT_WIDTH (currently fixed at 960)
+///     0x06-08 16bits: OUTPUT_HEIGHT (currently fixed at 720)
+/// TODO: configurable dimension and color depth support?
+/// TODO: behavior emulation closer to actual LCD display?
+pub(super) struct DisplayDevice {
+  vram: Vec<u8>,
+  frame_counter: i32,
+}
+
+const DISPLAY_WIDTH: u32 = 960;
+const DISPLAY_HEIGHT: u32 = 720;
+const REG_FLUSH: usize = 0x1FF0000;
+const REG_DIM: usize = 0x1FF0020;
+
+impl ShadowDevice for DisplayDevice {
+  fn new() -> Box<dyn ShadowDevice>
+  where
+    Self: Sized,
+  {
+    Box::new(Self {
+      vram: vec![0u8; (DISPLAY_WIDTH * DISPLAY_HEIGHT * 3) as usize],
+      frame_counter: 0,
+    })
+  }
+
+  fn read_mem(&self, addr: usize, size: usize) -> Vec<u8> {
+    let mut result = vec![0u8; size];
+    let start = addr;
+    let end = addr + size;
+
+    let vram_part_end = std::cmp::min(REG_FLUSH, end);
+    let vram_part_size = if vram_part_end > start {
+      vram_part_end - start
+    } else {
+      0
+    };
+
+    let reg_part_start = std::cmp::max(REG_FLUSH, start);
+    let reg_part_size = if end > reg_part_start {
+      end - reg_part_start
+    } else {
+      0
+    };
+
+    assert_eq!(
+      vram_part_size + reg_part_size,
+      size,
+      "Size calculation done wrong?"
+    );
+
+    self.vram_read(&mut result[0..vram_part_size], start, vram_part_size);
+    self.reg_read(
+      &mut result[vram_part_size..size],
+      reg_part_start - REG_FLUSH,
+      reg_part_size,
+    );
+
+    result
+  }
```

Comment:
It's too complicated. We don't need to support one AXI transaction touching both VRAM and controll registers.